File: packages/question-generation-service/src/question_generation_service/schemas/session.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field

from memosphere_domain import QuestionType

SessionStatus = Literal["active", "completed"]
FeedbackMode = Literal["immediate", "end", "never"]


class SessionQuestion(BaseModel):
    """Answer-stripped view of a stored question — no correct_answers, no
    explanation. Those only ever go out in AnswerResult, after the learner
    has submitted something to grade.
    """

    id: UUID
    question_type: QuestionType
    question_text: str
    options: list[str] | None
    estimated_time_seconds: int | None


class StartSessionRequest(BaseModel):
    """feedback_mode omitted → the user's remembered default; provided → used
    for this session and saved as the new default.
    """

    feedback_mode: FeedbackMode | None = None


class SessionState(BaseModel):
    session_id: UUID
    quiz_id: UUID | None
    feedback_mode: FeedbackMode
    total_questions: int
    current_index: int
    # None while the session is active outside immediate mode — running
    # correctness must not leak before the results page.
    correct_count: int | None
    session_complete: bool
    question: SessionQuestion | None


class SubmitAnswerRequest(BaseModel):
    question_id: UUID
    selected: list[str] = Field(min_length=1)
    response_time_seconds: int | None = None


class AnswerResult(BaseModel):
    """Grading fields are None outside immediate mode — the answer is
    recorded silently and only surfaces on the results page.
    """

    is_correct: bool | None
    correct_answers: list[str] | None
    explanation: str | None
    state: SessionState


class SessionSummary(BaseModel):
    """quiz_title/thema label the session even after its quiz is deleted;
    quiz_deleted marks the tombstone so clients can disable retake and badge
    the results.
    """

    session_id: UUID
    quiz_id: UUID | None
    quiz_title: str | None
    quiz_deleted: bool
    thema: str | None
    total_questions: int
    correct_answers: int
    accuracy: float
    total_time_seconds: int | None
    status: SessionStatus


class SessionReviewEntry(BaseModel):
    """correct_answers/explanation are None in never mode — an exam-mode
    session reveals per-question correctness but never the expected answer.
    is_correct is None for ungraded legacy responses.
    """

    question_id: UUID
    question_type: QuestionType
    question_text: str
    options: list[str] | None
    selected: list[str]
    is_correct: bool | None
    correct_answers: list[str] | None
    explanation: str | None


class SessionReview(BaseModel):
    summary: SessionSummary
    feedback_mode: FeedbackMode
    entries: list[SessionReviewEntry]


class SessionPreferences(BaseModel):
    default_feedback_mode: FeedbackMode


class SessionHistoryEntry(SessionSummary):
    feedback_mode: FeedbackMode
    started_at: datetime


class SessionHistory(BaseModel):
    sessions: list[SessionHistoryEntry]

File: apps/web-app/src/shared/types/session.ts
import type { QuestionType } from '@/types/thema';

export type SessionStatus = 'active' | 'completed';

export type FeedbackMode = 'immediate' | 'end' | 'never';

export interface SessionQuestion {
  id: string;
  question_type: QuestionType;
  question_text: string;
  options: string[] | null;
  estimated_time_seconds: number | null;
}

export interface SessionState {
  session_id: string;
  quiz_id: string | null;
  feedback_mode: FeedbackMode;
  total_questions: number;
  current_index: number;
  correct_count: number | null;
  session_complete: boolean;
  question: SessionQuestion | null;
}

export interface SubmitAnswerRequest {
  question_id: string;
  selected: string[];
  response_time_seconds?: number;
}

export interface AnswerResult {
  is_correct: boolean | null;
  correct_answers: string[] | null;
  explanation: string | null;
  state: SessionState;
}

export interface SessionSummary {
  session_id: string;
  quiz_id: string | null;
  quiz_title: string | null;
  quiz_deleted: boolean;
  thema: string | null;
  total_questions: number;
  correct_answers: number;
  accuracy: number;
  total_time_seconds: number | null;
  status: SessionStatus;
}

export interface SessionReviewEntry {
  question_id: string;
  question_type: QuestionType;
  question_text: string;
  options: string[] | null;
  selected: string[];
  is_correct: boolean | null;
  correct_answers: string[] | null;
  explanation: string | null;
}

export interface SessionReview {
  summary: SessionSummary;
  feedback_mode: FeedbackMode;
  entries: SessionReviewEntry[];
}

export interface SessionPreferences {
  default_feedback_mode: FeedbackMode;
}

export interface SessionHistoryEntry {
  session_id: string;
  quiz_id: string | null;
  quiz_title: string | null;
  quiz_deleted: boolean;
  thema: string | null;
  status: SessionStatus;
  feedback_mode: FeedbackMode;
  total_questions: number;
  correct_answers: number;
  accuracy: number;
  started_at: string;
  total_time_seconds: number | null;
}

export interface SessionHistory {
  sessions: SessionHistoryEntry[];
}

export const FEEDBACK_MODES: readonly {
  value: FeedbackMode;
  label: string;
  description: string;
}[] = [
  {
    value: 'immediate',
    label: 'Instant',
    description: 'See the answer after each question',
  },
  {
    value: 'end',
    label: 'At the end',
    description: 'Review everything once you finish',
  },
  {
    value: 'never',
    label: 'Exam mode',
    description: 'Score only — answers stay hidden',
  },
];

const SCORE_BANDS: readonly { min: number; className: string }[] = [
  { min: 90, className: 'text-green-500' },
  { min: 75, className: 'text-lime-500' },
  { min: 60, className: 'text-yellow-500' },
  { min: 45, className: 'text-amber-500' },
  { min: 30, className: 'text-orange-500' },
  { min: 0, className: 'text-red-500' },
];

export function scoreColorClass(percent: number): string {
  const band = SCORE_BANDS.find(b => percent >= b.min);
  return band ? band.className : 'text-red-500';
}
